### PR TITLE
[pvr][dialogs][Estuary][keymaps] Rework Guide window's navigation controls

### PR DESCRIPTION
**Pull request by ksooo**
Currently, controlling the EPG grid with a remote (or other input devices with limited buttons) is kind of inconvenient.

Examples:

* Goto first, last, current programme or a certain date of selected channel 
=> context menu or
=> cursor traveling to desired grid location

* Go 12 hours back/forward
=> cursor traveling to desired grid location

* Select another channel group 
=> cursor traveling to the group selector on top of the grid window 

* previous/next channel group
=> cursor traveling to the group selector on top of the grid window, plus selecting desired group 

* Goto to first/last channel of the grid
=> cursor traveling to the desired channel

* Goto playing channel
=> cursor traveling to the desired channel

Especially cursor traveling can be annoying if you got many channels and/or many programme entries in the grid.

This PR introduces
* a new (optional) EPG grid control dialog
* enhanced grid control context menu as a fallback if skin does not support the new EPG grid control dialog

The new grid control dialog can be activated using "long click enter", which should be available on most remotes. All other "simple klick" buttons available on limited remotes are already taken. I tested it (ofc) and found "long click enter" very handy. However, there is also a new key binding, so that users can use whatever key they want for this feature.

How does it look like?

**Please note that the shown button pngs are not final. I still could need a helping hand here! I'm definitely the wrong person when it comes to art work!**
 
screenshot004

screenshot001

@ronie mind doing a review for the skin changes?
@Jalle19 got some time for a code review?

### CODE REVIEW

**Comment by arnova** [comment]
Nice improvements but why 12 hour forward/backward and not 24 (one day) ? The latter makes more sense to me.

**Comment by ksooo** [comment]
> why 12 hour forward/backward and not 24 (one day) ?

 Was requested this way by users. Turned out to be more handy than 24 hours. We have this functionality for years in Kodi, before this PR it was accessible via remote buttons "next" and "previous".

File: xbmc/pvr/dialogs/GUIDialogPVRGuideControls.cpp
Comment (Rechi) [inline xbmc/pvr/dialogs/GUIDialogPVRGuideControls.cpp]: clang-format is currently set to add two spaces before `:`

File: xbmc/pvr/windows/GUIWindowPVRGuide.cpp
Comment (Rechi) [inline xbmc/pvr/windows/GUIWindowPVRGuide.cpp]: wrong include order, move one line up

**Comment by da-anda** [comment]
quite a lot of icons in the menu :). Might be tough to find somewhat self explaining icons for every of those features. My first thought was to maybe group them visually somehow (program, channel, group), but we don't have something like that elsewhere yet. Will have to think about it for some time.

**Comment by ksooo** [comment]
> My first thought was to maybe group them visually somehow (program, channel, group)
 
 currently, programme timeline controls are on the left, 
 channel and group controls are on the right

File: xbmc/pvr/dialogs/GUIDialogPVRGuideControls.cpp
Comment (ksooo) [reply xbmc/pvr/dialogs/GUIDialogPVRGuideControls.cpp]: fixed

File: xbmc/pvr/windows/GUIWindowPVRGuide.cpp
Comment (ksooo) [reply xbmc/pvr/windows/GUIWindowPVRGuide.cpp]: fixed

**Comment by da-anda** [comment]
> currently, programme timeline controls are on the left,
> channel and group controls are on the right

yes, seen that, but channels and group controls are still next to each other. My initial thought was if we could find a way to also separate those somehow, be it by some sort of group label, color schemes or some other way. But we don't do such things anywhere else yet.

Anyway, I try to come up with some icons if ronie (or anybody else) doesn't beat me to it and think about different approaches to the menu afterwards, so that my brainstorming doesn't block this PR

**Comment by da-anda** [comment]
I'm a bit rusted in how our interfaces work (JSON, python), so I thought I'd just ask. I've seen that you added those new shortcut methods to the PVRBuiltins - can those also be triggered via Python and JSON-RPC that way or would that require some extra work? And if the later is the case, would you mind exposing those to at least JSON-RPC as well? Thanks.

**Comment by ksooo** [comment]
> can those also be triggered via Python and JSON-RPC

 Sorry, no idea. Most probably extra work.

**Comment by Rechi** [comment]
jenkins build this please

File: xbmc/interfaces/builtins/PVRBuiltins.cpp
Comment (Jalle19) [inline xbmc/interfaces/builtins/PVRBuiltins.cpp]: An earily return would make sense here, that way you could save one `if (guideWindow)` check further down

File: xbmc/pvr/windows/GUIWindowPVRGuide.cpp
Comment (Jalle19) [inline xbmc/pvr/windows/GUIWindowPVRGuide.cpp]: If this is for v19, does it make sense to keep the fallback menu at all?

File: xbmc/pvr/windows/GUIWindowPVRGuide.cpp
Comment (Jalle19) [inline xbmc/pvr/windows/GUIWindowPVRGuide.cpp]: Aren't both of these the same as `CDateTimeSpan(0, std::abs(deltaHours), 0, 0)` now?

Comment (ksooo) [reply xbmc/pvr/windows/GUIWindowPVRGuide.cpp]: Haha, ofc. Good catch.

File: xbmc/interfaces/builtins/PVRBuiltins.cpp
Comment (ksooo) [reply xbmc/interfaces/builtins/PVRBuiltins.cpp]: yep, will change that

File: xbmc/pvr/windows/GUIWindowPVRGuide.cpp
Comment (ksooo) [reply xbmc/pvr/windows/GUIWindowPVRGuide.cpp]: I thought it would be nice not to enforce every skin to implement the new dialog...

**Comment by ksooo** [comment]
Thanks to @ronie we have a final layout for the new controls dialog:

screenshot000

Removing WIP label and will merge tomorrow, if nobody objects.

**Comment by da-anda** [comment]
thanks @ronie for stepping in with the icons and adding those group headers. I'm running in circles atm and unfortunately didn't find the time to work on the icons.

Looks good in general, but I'm not sure if we should really use icons that look like playback controls - might be a bit confusing even with the group headers. I was more thinking in icons in that direction (just a really quick mockup and not good looking at all)


Icon dummy

What do you guys think about this direction for icons?

**Comment by ksooo** [comment]
@da-anda I really like the icons you propose. Many thanks for sharing. If somebody hands me over the png files, I will be more than happy to incorporate them. We can do this at any time, even as follow-up for this PR.

**Comment by ronie** [comment]
@da-anda nice, the icons lgtm

**Review by ronie — APPROVED** [review]

**Comment by da-anda** [comment]
can probably fine tune the icons over the weekend. But feel free to merge the PR Kai and we do a follow up with the adjusted icons later

**Pull request merged**
